File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

long long solution(int n, vector<int> times) {
    long long l = 1;
    long long r = 1LL * (*max_element(times.begin(), times.end())) * n;
    
    while (l <= r) {
        long long m = (l + r) / 2;
        
        long long people = 0;
        for (int time : times) {
            people += m / time;
            if (people >= n) break;
        }
        
        if (people >= n) {
            r = m - 1;
        } else {
            l = m + 1;
        }
    }
    return l;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    answer = 0
    t_low, t_high = 1, max(times) * n   # 최소 심사 시간, 최대 심사 시간
    
    while t_low <= t_high:
        mid = (t_high + t_low) // 2
        people = 0
        for time in times:
            people += mid // time
            if people >= n:
                break
        if people >= n:
            answer = mid
            t_high = mid - 1
        else:
            t_low = mid + 1
            
    return answer